tutorials/stats/2_services_advanced.py

```diff
@@ -0,0 +1,103 @@
+# %% [markdown]
+"""
+# 2. Services Advanced
+
+The following examples shows how to decorate several functions
+for statistics collection.
+"""
+
+
+# %%
+import os
+import asyncio
+
+from dff.script import Context
+from dff.pipeline import Pipeline, ACTOR, Service, ExtraHandlerRuntimeInfo, to_service
+from dff.stats import StatsStorage, StatsExtractorPool, StatsRecord
+from dff.stats import default_extractor_pool  # import default pool from addon
+from dff.utils.testing.toy_script import TOY_SCRIPT
+from dff.utils.testing.common import is_interactive_mode
+
+
+# %% [markdown]
+"""
+As is the case with the regular handlers, you can add extractors
+both before and after the target service.
+You can use a handler that runs before the service to compare the pre-service and post-service
+states of the context, measure the running time, etc.
+An example of such handler can be found in the default extractor pool.
+
+Pass before- and after-handlers to the respective parameters of the `to_service` decorator.
+
+As for using multiple pools, you can subscribe your storage to any number of pools.
+
+"""
+
+
+# %%
+extractor_pool = StatsExtractorPool()
+
+
+@extractor_pool.add_extractor("after")
+async def get_service_state(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
+    # extract execution state of service from info
+    data = {
+        "execution_state": info["component"]["execution_state"],
+    }
+    # return a record to save into connected database
+    return StatsRecord.from_context(ctx, info, data)
+
+
+# %%
+# use together extractor_pool and default_extractor_pool
+# run extract_timing_before before `heavy_service`
+# run get_service_state and extract_timing_after after `heavy_service`
+@to_service(
+    before_handler=[default_extractor_pool["before"]["extract_timing"]],
+    after_handler=[get_service_state, default_extractor_pool["after"]["extract_timing"]],
+)
+async def heavy_service(ctx: Context):
+    _ = ctx  # get something from ctx if needed
+    await asyncio.sleep(0.02)
+
+
+# %%
+pipeline = Pipeline.from_dict(
+    {
+        "script": TOY_SCRIPT,
+        "start_label": ("greeting_flow", "start_node"),
+        "fallback_label": ("greeting_flow", "fallback_node"),
+        "components": [
+            Service(handler=heavy_service),  # add `heavy_service` before the actor
+            Service(
+                handler=to_service(
+                    before_handler=[default_extractor_pool["before"]["extract_timing"]],
```

Comment:
Wasn't the point of adding custom groups to allow doing so?
```
before_handler=extractor_pool["before"]  # <- a list of extractors
```
I don't understand why we add the ability to access extractors by names.
Seems like it can already be achieved by passing the function itself like so:
```
before_hanlder=[dff.stats.defaults.extract_timing_before]
```
So it feels like we should make the `__getitem__` function return a list of extractors in a specified group instead of a dict.

If the option to access individual extractors by name is needed then we should make a separate function to provide such access or modify `__getitem__` to accept keys in the format like `before.extract_timing` (which would return an extractor) along with the old format like `before` (which would return a list of extractors in the `before` group).

Reply:
I suggest that we make groups available as lists by overriding __getattr__, so that we don't mess with dictionary structures. Since we already provide `all_extractors` as a property, this would lead to better structure uniformity.